test: backend/tests/api/routes/test_ideas.py

cosmonewt opened the pull request:
added:
test_POST_create_idea_creates_idea_in_db
test_POST_create_idea_returns_422_when_invalid_data

Closes #353
Closes #354

cosmonewt committed:
test: backend/tests/api/routes/test_ideas.py
added:
test_POST_create_idea_creates_idea_in_db
test_POST_create_idea_returns_422_when_invalid_data

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -99,6 +99,27 @@ async def setup_downvote(real_db: AIOSession, idea: Idea, user: User):
     ),
 ]
 
+INVALID_IDEAS = [
+    pytest.param(
+        {
+            "name": "",
+            "description": "",
+        }
+    ),
+    pytest.param(
+        {
+            "name": "Test Name",
+            "description": "",
+        }
+    ),
+    pytest.param(
+        {
+            "name": "",
+            "description": "Test Description",
+        }
+    ),
```
Unless there's `id` argument added, to give distinct _name_ for a test case, wrapping with `pytest.param` is not needed.

Could you add additional cases for missing (not _just_ empty):
- `name`.
- `description`,
- both `name` and `description`?

cosmonewt replied:
Added test cases with missing (None) values.
Included ids for test names for more readability.

Ran the tests before updating the branch, and it seemed to work.

Appreciate the feedback, it is very insightful!

gikf replied:
I should be a bit more clear, by _missing_ I meant that the dict sent as json wouldn't have the necessary fields. 😅 

For example, if `name` is missing, the dict would be:
```python
{"description": "Test description"}
```

cosmonewt replied:
I see 😄. Made adjustments accordingly.

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -916,3 +937,55 @@ async def test_POST_create_idea_returns_idea(
+    user_with_client,
+    invalid_ideas,
+):
+    user, async_client = user_with_client
```
```suggestion
    _, async_client = user_with_client
```
If only second item of the unpacked `user_with_client` is used, the first one can be named with `_` to explicitly indicate it's not going to be used for anything.

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -916,3 +937,55 @@ async def test_POST_create_idea_returns_idea(
+@pytest.mark.integration
+@pytest.mark.anyio
+@pytest.mark.parametrize(
+    "invalid_ideas",
```
```suggestion
    "invalid_idea",
```
Nitpick - only one entry is used per individual test run.

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -916,3 +937,55 @@ async def test_POST_create_idea_returns_idea(
+async def test_POST_create_idea_returns_422_when_invalid_data(
+    real_db: AIOSession,
+    user_with_client,
+    invalid_ideas,
```
```suggestion
    invalid_idea,
```

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -916,3 +937,55 @@ async def test_POST_create_idea_returns_idea(
+    user, async_client = user_with_client
+
+    async with clean_new_ideas(real_db):
+        response = await async_client.post("/ideas/", json=invalid_ideas)
```
```suggestion
        response = await async_client.post("/ideas/", json=invalid_idea)
```

gikf commented on backend/tests/api/routes/test_ideas.py:
```diff
@@ -916,3 +937,55 @@ async def test_POST_create_idea_returns_idea(
+        assert response.status_code == 200
+
+        db_idea = await real_db.find_one(Idea, Idea.id == ObjectId(data["id"]))
+
```
```suggestion

assert db_idea is not None
```
A bit nitpicky, to make type checker stop freaking out, because `db_idea` can be `None`, if no entry was found. And asserts below _might_ be checking specified attributes on `None`.

cosmonewt committed:
Update backend/tests/api/routes/test_ideas.py

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

cosmonewt committed:
Update backend/tests/api/routes/test_ideas.py

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

cosmonewt committed:
Update backend/tests/api/routes/test_ideas.py

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

cosmonewt committed:
Update backend/tests/api/routes/test_ideas.py

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

cosmonewt committed:
Update backend/tests/api/routes/test_ideas.py

Co-authored-by: Krzysztof G. <60067306+gikf@users.noreply.github.com>

cosmonewt committed:
refactor: revised create idea test
- added test cases with missing inputs
- added test ids to pytest.param for readability

cosmonewt committed:
refactor: test cases
changed tests cases in variable INVALID_IDEAS from e.g.: "name": None, to omitting the key-value pair entirely
